src/emit/substrate/mod.rs

```diff
@@ -1851,3 +1853,47 @@ fn event_id<'b>(
 
     Some(binary.context.i8_type().const_int(event_id as u64, false))
 }
+
+/// Print the return code of API calls to the debug buffer.
+fn log_return_code<'b>(binary: &Binary<'b>, api: &'static str, code: &IntValue) {
+    if !binary.log_api_return_codes {
+        return;
+    }
+
+    emit_context!(binary);
+
+    let fmt = format!("{}=", api);
+    let msg = fmt.as_bytes();
+    let length = i32_const!(msg.len() as u64 + 16);
+    let out_buf = binary.vector_new(length, i32_const!(1), None);
+    let out_buf_ptr = binary.vector_bytes(out_buf.into());
```

Comment:
Rather than heap allocating this, just use `builder.build_array_alloca()`. Faster and less wastage.